shared/src/main/scala/mlscript/NewLexer.scala

```diff
@@ -31,6 +31,9 @@ class NewLexer(origin: Origin, raise: Diagnostic => Unit, dbg: Bool) {
     c === '0' || c === '1'
   def isDigit(c: Char): Bool =
     c >= '0' && c <= '9'
+  def matches(i: Int, syntax: Str): Bool =
+    i + syntax.length <= length && bytes.slice(i, i + syntax.length).mkString === syntax
```

Comment:
This is very inefficient. Why not a tail-recursive `matches(i: Int, syntax: Str, start: Int)` which checks each character until `start >= syntax.length`?